File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        letter_dict = {}
        result = 0
        start = 0
        for i in range(len(s)):
            idx = letter_dict.get(s[i])
            if idx is None or idx < start:
                if i - start + 1 > result:
                    result = i - start + 1
            else:
                start = idx + 1
            letter_dict[s[i]] = i
        
        return result
            

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    letterDict := map[uint8]int{}
    result := 0
    start := 0
    for i := 0; i < len(s); i++ {
        idx, ok := letterDict[s[i]]
        if !ok || idx < start {
            if i-start+1 > result {
                result += 1
            }
        } else {
            start = idx + 1
        }
        
        letterDict[s[i]] = i
    }
    
    return result
}